WIP: added app context manager (open and close methods) (#270)
peopledoc/procrastinate: docs/howto/app.rst

        connector=connector,
        import_paths=["dotted.path.to", "all.the.modules", "that.define.tasks"]
    )

> 

Reviewer: The title of the page is not very relevant anymore.

Author: Do you mean changing 

> Find a correct place for your app

?

Reviewer: yes

Author: I changed a few files in the doc to improve it. The open/close explanation has now its own dedicated file : `open_connection.rst`, which is the third step in the Howto : 
![image](https://user-images.githubusercontent.com/16255570/90871471-6dcfea80-e39b-11ea-9490-4dc4e38db484.png)

If this is good, we might be ready to merge :+1:

Author: I think it deserves its own file, as it not directly related to connector creation, nor App creation, and must be done before defining tasks.

Initiate and terminate the connection to the database
-----------------------------------------------------

The app must be opened at the beginning of the program to create the connection pool to
the database, and eventually closed to properly terminate it.

The app implements several ways to do this.

Explicitly open and close the app
^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^

The app can be opened using the ``open`` method (for sync case) or ``open_async``
method (for async case). It eventually needs to be closed, using the ``close`` method
(for sync case) or ``close_async`` (for async case).

For sync case::

    app.open()
    ...
    app.close()

For async case::

    await app.open_async()
    ...
    await app.close_async()

> Open the app and automatically close it with a context manager
> ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^

Reviewer: ```suggestion
Open and close the app with a context manager
^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
```

The app can be opened with a context manager, which will automatically close it when
leaving the context.

For sync case::

    with app.open():
        pass

For async case::

    async with app.open_async():
        pass

Create and open the app in a single line
^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^

Both ``open`` and ``open_async`` return the current app when they are called or
entered, which allows you to instantiate and open your app in a single line.

For sync case::

>    app = procrastinate.App(connector=connector).open()

Author: Ok, I delete the three others

Or with a context manager::

   with procrastinate.App(connector=connector).open() as app:
       pass

For async case::

   app = await procrastinate.App(connector=connector).open_async()

Or with a context manager::

   async with procrastinate.App(connector=connector).open_async() as app:
       pass

Open the app, and let the garbage collector terminate the connection
^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^

> The app does not necessarily need to be closed at the end of your program. Indeed, the

Reviewer: ```suggestion
The app does not necessarily need to be closed at the end of your program. The
```

connections to the database will be automatically terminated when the database
connector is garbage collected. Thus, your program can work with only the following
statements, as we did in the readme and the quickstart::

   app = procrastinate.App(connector=connector)
   app.open()

> 

Reviewer: ```suggestion

But remember: "explicit is better than implicit".
```

        connector=connector,
        import_paths=["dotted.path.to", "all.the.modules", "that.define.tasks"]
    )

Initiate and terminate the connection to the database
-----------------------------------------------------

The app must be opened at the beginning of the program to initiate the connection to
the database, and eventually closed to properly terminate it.

The app implements several ways to do this.

Explicitly open and close the app
^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^

The app can be opened using the ``open`` method (for sync case) or ``open_async``

> method (for async case). It eventually needs to be close, using the ``close`` method

Reviewer 2: ```suggestion
method (for async case). It eventually needs to be closed, using the ``close`` method
```

(for sync case) or ``close_async`` (for async case).

For sync case::

    app.open()
    ...
    app.close()

For async case::

    await app.open_async()
    ...
    await app.close_async()

Open the app and automatically close it with a context manager
^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^

The app can be opened with a context manager, which will automatically close it when
leaving the context.

For sync case::

    with app.open():
        pass

For async case::

    async with app.open_async():
        pass

Create and open the app in a single line
^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^

> Both ``open`` and ``open_async`` return the current app when they are called or

Reviewer 2: Strictly speaking `open_async` returns an awaitable, which `__await__` method returns the application. Do we want to be more accurate here in the documentation?

Author: You are right. I tried to keep this as simple as possible. Is the following better ?
```suggestion
Both ``open`` and ``open_async`` return an ``AwaitableContext`` object that returns the current app when it is entered, or awaited
```
It is more accurate, but I am afraid it improves complexity too much.

Reviewer: Yep, I think it's counter-productive to mention the AwaitableContext here. I'm comfortable with the original sentence.

entered, which allows you to instantiate and open your app in a single line.

For sync case::

   app = procrastinate.App(connector=connector).open()

Or with a context manager::

   with procrastinate.App(connector=connector).open() as app:
       pass

For async case::

   app = await procrastinate.App(connector=connector).open_async()

Or with a context manager::

   async with procrastinate.App(connector=connector).open_async() as app:
       pass

Open the app, and let the garbage collector terminate the connection
^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^

> The app does not necessarily needs to be closed at the end of your program. Indeed, the

Reviewer 2: ```suggestion
The app does not necessarily need to be closed at the end of your program. Indeed, the
```

connections to the database will be automatically terminated when the database

> connectors are garbage collected. Thus, your program can work with only the following

Reviewer 2: ```suggestion
connector is garbage collected. Thus, your program can work with only the following
```

> statements, as we did in the quickstart::

Reviewer 2: In the `README.rst` file `app.close()` is called, which might not be necessary either.

Author: It is not necessary indeed, but I think it adds clarity. If `app.close()` was not mentioned in the README.rst, the user might wonder why we do not close the connection.

Author: I remove `app.close()` to keep this example as light as possible.

Reviewer 2: :+1:

   app = procrastinate.App(connector=connector)
   app.open()

> Note that the ``close`` method is not called in he above example.

Reviewer 2: ```suggestion
Note that the ``close`` method is not called in the above example.
```

> Note that the ``close`` method is not called in the above example.

Reviewer 2: I think this sentence is superfluous.

Author: Totally agree ! I change that :ok_hand:
